### PR TITLE
CleveresTricky 2.6.0: Rust-first runtime, hardened attestation & Vault UX

### PR DESCRIPTION
**Pull request by tryigit**
# CleveresTricky 2.6.0

This is a large architectural release rather than a small feature update. The Android-facing behavior stays familiar, while most portable parsing, crypto, attestation rewriting, keybox handling and service transport now live behind a hardened Rust-first runtime.

## What users should notice

- More reliable module startup and recovery across the daemon, Android adapter, WebUI bridge and backend.
- WebUI Runtime Health and file operations no longer depend on a fragile startup race.
- Keybox activation is fail-closed and preserves the previous healthy snapshot when a replacement cannot be activated.
- The companion Encryptor/Vault app has a redesigned monochrome full-screen UX, in-app language selection, earlier keybox validation, clearer errors and less UI-thread I/O.
- Final module packaging now checks runtime payload presence, installer/supervisor wiring, checksums and abnormal payload shrinkage so accidental omissions fail CI.

## Timing regression fixed from device testing

Real-device testing exposed a 2.6.0 timing regression around attested vs ordinary KeyMint operations. The problem was not hidden by changing the detector threshold or adding artificial delays.

The hot path was corrected instead:

- ordinary/non-attested `generateKey` replies are classified locally from the already-parsed X.509 leaf and exit before Rust certificate-backend IPC;
- genuine issuer-chain parsing is deferred until a real rewrite is actually needed;
- targeted `getKeyEntry` readback follows the same local attestation classification/cached-replacement strategy, avoiding a non-attested-only backend inspection round trip;
- dedicated regression tests lock both creation and readback fast-path ordering so backend work cannot silently move ahead of the local guard again.

This removes the asymmetric UDS/parser work that had reappeared compared with the 2.5.8 behavior while keeping Rust-backed certificate rewriting for actual attested keys.

## Architecture

The privileged process is now a minimal supervisor/capability broker. Complex parsing and cryptographic work runs in an unprivileged Rust backend. The design includes bounded authenticated local IPC, descriptor-relative secure filesystem access, validated FD transfer, opaque backend key IDs, bounded WebUI staging, restart identity/stale-handle invalidation and controlled state recovery.

Portable CBOX/backup crypto, XML/keybox parsing, Android-attestation/X.509 rewriting, CRL processing and portable policy parsing moved behind Rust boundaries. The existing private-libbinder C++ ABI island remains narrow and was not expanded.

## Runtime fixes found on real devices

Device testing found failures that host CI had not reproduced. The branch fixes Android adapter publication/startup ordering, removes a fragile `/proc/<pid>/exe` trust dependency, whitelists only the required WebUI staging capabilities, fixes bounded staged-response streaming semantics and starts diagnostics early enough to remain reachable when the heavier backend is unhealthy.

CI now contains generalized runtime wiring and artifact-integrity contracts for these classes of failures rather than tests tied only to one historical bug.

## Validation

Current merge head: `72adb82d2020393b1204a6acc0702c2da9c2f46b`

Completed successfully on this exact head:

- Security Regression `32148052612`
- Native Hardening `32148051960`
- Rust Lockfile Check `32148052271`
- Rust Fuzz Smoke `32148052135`
- Build Quality/Lint gate
- Rust fmt, Clippy, workspace tests and `cargo audit`
- Android-target Rust lint
- JVM unit tests, including timing fast-path regressions
- module artifact build

Build run `32148052349` was still finishing the companion-app/final archive tail when merge was explicitly authorized; its core code, test and module-build gates had already passed.

`CHANGELOG.md` is updated for 2.6.0 in user-facing English, focusing on practical benefits while still making clear that this release replaces a substantial part of the internal runtime architecture.

## Performance note

Physical-device PSS/RSS and end-to-end Binder timing are not fabricated from CI. The device benchmark remains the final performance signal. The specific asymmetric backend work identified from the reported timing warning has been removed structurally and is protected by regression tests.

### CODE REVIEW

File: rust/crypto-core/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+    Ok(bytes)
+}
+
+pub fn encrypt_backup(plaintext: &[u8], password: &str) -> Result<Vec<u8>, CryptoError> {
+    if plaintext.len() > MAX_BACKUP_PLAINTEXT_BYTES
+        || utf16_units(password) > MAX_PASSWORD_UTF16_UNITS
+    {
+        return Err(CryptoError::InvalidInput);
+    }
+    let mut salt = [0u8; SALT_BYTES];
```
[the remark this answers is not in the document]
Comment (tryigit) [reply rust/crypto-core/src/lib.rs]: Decoded as a false positive. This is only zero-initialization of an output buffer for `getrandom`; the buffer is cryptographically filled before it is passed to PBKDF2/AES-GCM. If `getrandom` fails, the function zeroizes both buffers and returns `RandomUnavailable` before either value can be used. I am keeping the fail-closed control flow and will resolve this thread after the corrected head passes CI.

File: rust/crypto-core/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+    let mut iv = [0u8; IV_BYTES];
```
[the remark this answers is not in the document]
Comment (tryigit) [reply rust/crypto-core/src/lib.rs]: Decoded as a false positive for the same reason as the salt finding: this is a zero-initialized destination buffer that `getrandom` fills before AES-GCM ever sees it. RNG failure returns early after zeroization, so a zero IV cannot reach encryption. I will resolve after CI confirms the corrected head.

File: rust/crypto-core/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        .len()
+        .checked_sub(TAG_BYTES)
+        .ok_or(CryptoError::InvalidInput)?;
+    if body_end < HEADER_BYTES {
+        return Err(CryptoError::InvalidInput);
+    }
+
+    let mut header = [0u8; HEADER_BYTES];
+    header.copy_from_slice(&bytes[..HEADER_BYTES]);
+    let mut salt = [0u8; SALT_BYTES];
```
[the remark this answers is not in the document]
Comment (tryigit) [reply rust/crypto-core/src/lib.rs]: Decoded as a false positive. This zero-initialized array is a temporary parse buffer on the decrypt path; it is immediately overwritten from the authenticated envelope header with `copy_from_slice` before PBKDF2 uses it. It is also zeroized on all exits. No hard-coded salt participates in decryption.

File: rust/crypto-core/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+fn utf16_units(value: &str) -> usize {
+    value.encode_utf16().count()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use base64::Engine;
+
+    const PASSWORD: &str = "correct horse battery staple";
```
[the remark this answers is not in the document]
Comment (tryigit) [reply rust/crypto-core/src/lib.rs]: Decoded as a test-only deterministic compatibility vector. `PASSWORD` is inside `#[cfg(test)]` and exists solely so the Rust implementation can reproduce CTSB/CBOX golden ciphertext generated by the managed oracle; it is never compiled into production artifacts and is not a credential. I’ll keep the deterministic vector for differential testing and resolve the thread after the corrected head passes CI.

File: rust/backend/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+    )?;
+    stream.flush()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use base64::Engine as _;
+
+    const PASSWORD: &str = "correct horse battery staple";
```
[the remark this answers is not in the document]
Comment (tryigit) [reply rust/backend/src/main.rs]: Decoded as a test-only deterministic compatibility vector, not a production credential. This `PASSWORD` is inside the backend's `#[cfg(test)]` module and is used only to decrypt the fixed CTSB/CBOX oracle ciphertext and exercise backend framing/round-trip behavior. Rust excludes this module from release builds. The production backend receives passwords only from bounded IPC, never from this constant. I’ll resolve the thread after the corrected backend head passes host + Android validation.

File: rust/backend/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+    #[test]
+    fn backup_encrypt_accepts_empty_password_and_plaintext() {
+        let encrypted = handle_request(
+            OP_CRYPTO_BACKUP_ENCRYPT,
+            encode_backup_request("", b""),
+        )
+        .unwrap();
+        let decrypted = handle_request(
+            OP_CRYPTO_BACKUP_DECRYPT,
+            encode_backup_request("", &encrypted),
```
[the remark this answers is not in the document]
Comment (tryigit) [reply rust/backend/src/main.rs]: Confirmed on fully green head `6d4bc34`: this reference is inside `rust/backend/src/main.rs`'s `#[cfg(test)]` module. The deterministic password exists only to reproduce the fixed CTSB/CBOX compatibility vector; release builds do not compile it and production passwords arrive only through bounded IPC. Host fmt/Clippy/tests/audit, both Android Rust targets, JVM tests, release ZIP and ELF hardening all pass.

File: rust/backend/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+        )
+        .unwrap();
+        assert!(decrypted.is_empty());
+    }
+
+    #[test]
+    fn cbox_prefix_accepts_empty_password_like_managed_oracle() {
+        let request = encode_cbox_request("", "", b"body");
```
[the remark this answers is not in the document]
Comment (tryigit) [reply rust/backend/src/main.rs]: Confirmed on fully green head `6d4bc34`: this use is in the backend `#[cfg(test)]` module and references the same deterministic compatibility password. It is excluded from release artifacts; the production backend has no hard-coded password path. Full host, Android-target, JVM, module archive and ELF-hardening validation passed.

File: rust/backend/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+            encode_backup_request(PASSWORD, &encrypted),
+        )
+        .unwrap();
+        assert_eq!(decrypted, plaintext);
+    }
+
+    #[test]
+    fn backup_encrypt_accepts_empty_password_and_plaintext() {
+        let encrypted =
+            handle_request(OP_CRYPTO_BACKUP_ENCRYPT, encode_backup_request("", b"")).unwrap();
```
[the remark this answers is not in the document]
Comment (tryigit) [reply rust/backend/src/main.rs]: Confirmed on fully green head `6d4bc34`: this is also test-only code below `#[cfg(test)]`, used solely for empty-password/CBOX compatibility coverage. It is not a production credential and is not compiled into release builds. The corrected head passed Security Regression, host Rust checks/audit, both Android Rust targets, JVM tests, release ZIP verification and ELF hardening.

**Comment by tryigit** [comment]
Final post-fix validation completed on head `ad90e82673df37e2366831ea356811aeb25bcfc5` (tree `e573e5650439cef67e715451ca26c1dac799843e`).

All final PR workflows are green:
- Build `32085366332`: success — Quality Gate/Lint, Rust fmt/Clippy/tests/audit, Android Rust lint, JVM tests, module artifact, Encryptor app, archive/checksum validation and native hardening all passed.
- Security Regression `32085366296`: success.
- Native Hardening `32085366341`: success.
- Rust Lockfile Check `32085366353`: success.
- Rust Fuzz Smoke `32085366323`: success.

The activation-failure/keybox regression and Android Indonesian locale/lint issues found during final validation are fixed and covered by the passing suites. All inline review threads are resolved/outdated. PR #953 has been marked Ready for review. No merge or auto-merge action was performed.

**Comment by tryigit** [comment]
Runtime repair checkpoint — PR remains **Draft** and is **not ready for merge** until device validation.

Current validated source head: `6278848d3ae259bd8cd5916f99ffc940a9df6d1a`.

Device-root causes fixed in this repair pass:
- Android secure-file broker no longer depends on `/proc/<daemon>/exe`; it authenticates the direct root daemon parent by peer credentials/PID.
- Initial config capability handshake now tolerates the bounded daemon→adapter PID-publication race.
- Broker whitelist now explicitly supports required `webui_bridge/` and `webui_bridge/staging/` directories; previously the adapter could exit before WebUI registration because only `keyboxes/` mkdir was allowed.
- Service-owned `webui_bridge/staging/<32hex>.download` responses now use bounded 64 KiB create/append staging with a 20 MiB aggregate cap. The old exact-length interpretation of `SecureFile.writeStream(limit)` could fail normal large-response staging.
- Upload staging remains owned by the separate Rust `webui_bridge` process; daemon capability surface was not broadened to `.upload`.

General regression prevention added:
- runtime wiring contract in the JVM security suite (supervisor→daemon→adapter→backend, WebUI staging, backend auth/UID, always-on Keystore hook activation order),
- Rust broker whitelist/chunk/traversal/fail-closed tests,
- module build payload-integrity floors and installer/supervisor wiring checks,
- Security Regression failure-report artifacts.

Current-head CI: Build `32098702732`, Security Regression `32098702736`, Native Hardening `32098702774`, Rust Lockfile `32098702731`, Rust Fuzz `32098702742` — all **success**.

Build artifact id `9311103227`: `CleveresTricky-V2.6.0-2820-e7e8e025-release.zip`, 2,904,023 bytes, SHA-256 `894ef77cadeacb28dd82c7378b2c1936126ff04b18c42201dbf9d82e20d496a2`. Independent archive audit found no missing/orphan/mismatched checksum, duplicate/traversal/symlink entry, or missing mandatory runtime payload. The PR stays Draft pending a real KernelSU/APatch device install/reboot test of WebUI Runtime Health and Key Attestation interception.
